src/devices/sound/discrete.cpp

```diff
@@ -204,74 +209,78 @@ inline void discrete_task::step_nodes()
 		}
 	}
 
-	/* buffer the outputs */
+	// buffer the outputs
+	for (output_buffer &outbuf : m_buffers)
+		*outbuf.ptr.load(std::memory_order_relaxed) = *outbuf.source;
+	std::atomic_thread_fence(std::memory_order_release);
 	for (output_buffer &outbuf : m_buffers)
-		*outbuf.ptr++ = *outbuf.source;
+		outbuf.ptr.store(outbuf.ptr.load(std::memory_order_relaxed) + 1, std::memory_order_relaxed);
 }
 
 void *discrete_task::task_callback(void *param, int threadid)
 {
 	const auto &list = *reinterpret_cast<const discrete_sound_device::task_list_t *>(param);
-	do
+	while (true)
 	{
 		for (const auto &task : list)
 		{
-			/* try to lock */
+			// try to lock
 			if (task->lock_threadid(threadid))
 			{
 				if (!task->process())
 					return nullptr;
 				task->unlock();
 			}
 		}
-	} while (1);
+	}
 
 	return nullptr;
 }
 
-bool discrete_task::process()
+inline bool discrete_task::process()
 {
-	int samples = std::min(int(m_samples), MAX_SAMPLES_PER_TASK_SLICE);
+	int samples = std::min(m_samples, MAX_SAMPLES_PER_TASK_SLICE);
 
-	/* check dependencies */
+	// check dependencies
 	for (input_buffer &sn : source_list)
 	{
-		int avail = sn.linked_outbuf->ptr - sn.ptr;
+		const int avail = sn.linked_outbuf->ptr.load(std::memory_order_relaxed) - sn.ptr;
 		if (avail < 0)
 			throw emu_fatalerror("discrete_task::process: available samples are negative");
 		if (avail < samples)
 			samples = avail;
 	}
+	std::atomic_thread_fence(std::memory_order_acquire);
 
 	m_samples -= samples;
 	if (m_samples < 0)
 		throw emu_fatalerror("discrete_task::process: m_samples got negative");
 	while (samples > 0)
 	{
-		/* step */
+		// step
 		step_nodes();
 		samples--;
 	}
 	if (m_samples == 0)
 	{
-		/* return and keep the task locked so it is not picked up by other worker threads */
+		// return and keep the task locked so it is not picked up by other worker threads
 		return false;
 	}
 	return true;
 }
 
 void discrete_task::prepare_for_queue(int samples)
 {
+	m_threadid.store(-1, std::memory_order_relaxed); // unlock the thread
 	m_samples = samples;
-	/* set up task buffers */
+
+	// set up task buffers
 	for (output_buffer &ob : m_buffers)
-		ob.ptr = ob.node_buf.get();
+		ob.ptr.store(ob.node_buf.get(), std::memory_order_relaxed);
 
-	/* initialize sources */
+	// initialize sources
 	for (input_buffer &sn : source_list)
-	{
 		sn.ptr = sn.linked_outbuf->node_buf.get();
-	}
 }
```

Comment:
We can use relaxed memory order here because there’s a barrier after preparing all the tasks.